sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/joins.scala

```diff
@@ -253,12 +253,28 @@ trait JoinSelectionHelper {
     val buildLeft = if (hintOnly) {
       hintToShuffleHashJoinLeft(hint)
     } else {
-      canBuildLocalHashMapBySize(left, conf) && muchSmaller(left, right)
+      if (hintToPreferShuffleHashJoinLeft(hint)) {
+        true
+      } else {
+        if (!conf.preferSortMergeJoin) {
+          canBuildLocalHashMapBySize(left, conf) && muchSmaller(left, right)
+        } else {
+          false
+        }
+      }
     }
     val buildRight = if (hintOnly) {
       hintToShuffleHashJoinRight(hint)
     } else {
-      canBuildLocalHashMapBySize(right, conf) && muchSmaller(right, left)
+      if (hintToPreferShuffleHashJoinRight(hint)) {
+        true
+      } else {
+        if (!conf.preferSortMergeJoin) {
```

Comment:
is this change related to this PR? the old code was `canBuildLocalHashMapBySize(right, conf) && muchSmaller(right, left)`